In [67]:
import pandas as pd
url="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_aa04e98df66b49989ab49f9f2ac12237.csv"
df=pd.read_csv(url)
df.head()

Age  Weight  Length    Sex        BMI  DM  HTN  Current Smoker  EX-Smoker  \
0   53      90     175   Male  29.387755   0    1               1          0   
1   67      70     157  Fmale  28.398718   0    1               0          0   
2   54      54     164   Male  20.077335   0    0               1          0   
3   66      67     158  Fmale  26.838648   0    1               0          0   
4   50      87     153  Fmale  37.165193   0    1               0          0   

   FH  ... Lymph Neut  PLT EF-TTE Region RWMA     VHD       LAD       LCX  \
0   0  ...    39   52  261     50           0       N  Stenotic    Normal   
1   0  ...    38   55  165     40           4       N  Stenotic  Stenotic   
2   0  ...    38   60  230     40           2    mild  Stenotic    Normal   
3   0  ...    18   72  742     55           0  Severe    Normal    Normal   
4   0  ...    55   39  274     50           0  Severe    Normal    Normal   

        RCA    Cath  
0  Stenotic     CAD  
1    Normal     CAD  
2    Normal     CAD  
3    Normal  Normal  
4    Normal  Normal  

[5 rows x 59 columns]

In [68]:
df.shape

(303, 59)

In [69]:
y=df[["LAD"]]
y.groupby("LAD")["LAD"].count()

LAD
Normal      126
Stenotic    177
Name: LAD, dtype: int64

In [70]:
X=df.drop(["LAD","LCX","RCA","Cath"],axis=1)
X.head()

Age  Weight  Length    Sex        BMI  DM  HTN  Current Smoker  EX-Smoker  \
0   53      90     175   Male  29.387755   0    1               1          0   
1   67      70     157  Fmale  28.398718   0    1               0          0   
2   54      54     164   Male  20.077335   0    0               1          0   
3   66      67     158  Fmale  26.838648   0    1               0          0   
4   50      87     153  Fmale  37.165193   0    1               0          0   

   FH  ...    HB    K   Na    WBC Lymph Neut  PLT  EF-TTE  Region RWMA     VHD  
0   0  ...  15.6  4.7  141   5700    39   52  261      50            0       N  
1   0  ...  13.9  4.7  156   7700    38   55  165      40            4       N  
2   0  ...  13.5  4.7  139   7400    38   60  230      40            2    mild  
3   0  ...  12.1  4.4  142  13000    18   72  742      55            0  Severe  
4   0  ...  13.2  4.0  140   9200    55   39  274      50            0  Severe  

[5 rows x 55 columns]

In [71]:
types=pd.DataFrame(X.dtypes,columns=["type"])
types

type
Age                      int64
Weight                   int64
Length                   int64
Sex                     object
BMI                    float64
DM                       int64
HTN                      int64
Current Smoker           int64
EX-Smoker                int64
FH                       int64
Obesity                 object
CRF                     object
CVA                     object
Airway disease          object
Thyroid Disease         object
CHF                     object
DLP                     object
BP                       int64
PR                       int64
Edema                    int64
Weak Peripheral Pulse   object
Lung rales              object
Systolic Murmur         object
Diastolic Murmur        object
Typical Chest Pain       int64
Dyspnea                 object
Function Class           int64
Atypical                object
Nonanginal              object
Exertional CP           object
LowTH Ang               object
Q Wave                   int64
St Elevation             int64
St Depression            int64
Tinversion               int64
LVH                     object
Poor R Progression      object
BBB                     object
FBS                      int64
CR                     float64
TG                       int64
LDL                      int64
HDL                    float64
BUN                      int64
ESR                      int64
HB                     float64
K                      float64
Na                       int64
WBC                      int64
Lymph                    int64
Neut                     int64
PLT                      int64
EF-TTE                   int64
Region RWMA              int64
VHD                     object

In [72]:
objects=types[types["type"]=="object"]
objects

type
Sex                    object
Obesity                object
CRF                    object
CVA                    object
Airway disease         object
Thyroid Disease        object
CHF                    object
DLP                    object
Weak Peripheral Pulse  object
Lung rales             object
Systolic Murmur        object
Diastolic Murmur       object
Dyspnea                object
Atypical               object
Nonanginal             object
Exertional CP          object
LowTH Ang              object
LVH                    object
Poor R Progression     object
BBB                    object
VHD                    object

In [73]:
object_index=objects.index.tolist()
object_index

['Sex',
 'Obesity',
 'CRF',
 'CVA',
 'Airway disease',
 'Thyroid Disease',
 'CHF',
 'DLP',
 'Weak Peripheral Pulse',
 'Lung rales',
 'Systolic Murmur',
 'Diastolic Murmur',
 'Dyspnea',
 'Atypical',
 'Nonanginal',
 'Exertional CP',
 'LowTH Ang',
 'LVH',
 'Poor R Progression',
 'BBB',
 'VHD']

In [74]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
objects=X[object_index]
objects=objects.apply(le.fit_transform)
objects.head()

Sex  Obesity  CRF  CVA  Airway disease  Thyroid Disease  CHF  DLP  \
0    1        1    0    0               0                0    0    1   
1    0        1    0    0               0                0    0    0   
2    1        0    0    0               0                0    0    0   
3    0        1    0    0               0                0    0    0   
4    0        1    0    0               0                0    0    0   

   Weak Peripheral Pulse  Lung rales  ...  Diastolic Murmur  Dyspnea  \
0                      0           0  ...                 0        0   
1                      0           0  ...                 0        0   
2                      0           0  ...                 0        0   
3                      0           0  ...                 1        1   
4                      0           0  ...                 0        1   

   Atypical  Nonanginal  Exertional CP  LowTH Ang  LVH  Poor R Progression  \
0         0           0              0          0    0                   0   
1         0           0              0          0    0                   0   
2         0           0              0          0    0                   0   
3         0           1              0          0    0                   0   
4         0           0              0          0    0                   0   

   BBB  VHD  
0    1    1  
1    1    1  
2    1    3  
3    1    2  
4    1    2  

[5 rows x 21 columns]

In [75]:
numeric=types[types["type"]!="object"]
numeric

type
Age                   int64
Weight                int64
Length                int64
BMI                 float64
DM                    int64
HTN                   int64
Current Smoker        int64
EX-Smoker             int64
FH                    int64
BP                    int64
PR                    int64
Edema                 int64
Typical Chest Pain    int64
Function Class        int64
Q Wave                int64
St Elevation          int64
St Depression         int64
Tinversion            int64
FBS                   int64
CR                  float64
TG                    int64
LDL                   int64
HDL                 float64
BUN                   int64
ESR                   int64
HB                  float64
K                   float64
Na                    int64
WBC                   int64
Lymph                 int64
Neut                  int64
PLT                   int64
EF-TTE                int64
Region RWMA           int64

In [76]:
numeric_index=numeric.index.tolist()
numeric_index

['Age',
 'Weight',
 'Length',
 'BMI',
 'DM',
 'HTN',
 'Current Smoker',
 'EX-Smoker',
 'FH',
 'BP',
 'PR',
 'Edema',
 'Typical Chest Pain',
 'Function Class',
 'Q Wave',
 'St Elevation',
 'St Depression',
 'Tinversion',
 'FBS',
 'CR',
 'TG',
 'LDL',
 'HDL',
 'BUN',
 'ESR',
 'HB',
 'K',
 'Na',
 'WBC',
 'Lymph',
 'Neut',
 'PLT',
 'EF-TTE',
 'Region RWMA']

In [77]:
numeric=X[numeric_index]
numeric.head()

Age  Weight  Length        BMI  DM  HTN  Current Smoker  EX-Smoker  FH  \
0   53      90     175  29.387755   0    1               1          0   0   
1   67      70     157  28.398718   0    1               0          0   0   
2   54      54     164  20.077335   0    0               1          0   0   
3   66      67     158  26.838648   0    1               0          0   0   
4   50      87     153  37.165193   0    1               0          0   0   

    BP  ...  ESR    HB    K   Na    WBC  Lymph  Neut  PLT  EF-TTE  Region RWMA  
0  110  ...    7  15.6  4.7  141   5700     39    52  261      50            0  
1  140  ...   26  13.9  4.7  156   7700     38    55  165      40            4  
2  100  ...   10  13.5  4.7  139   7400     38    60  230      40            2  
3  100  ...   76  12.1  4.4  142  13000     18    72  742      55            0  
4  110  ...   27  13.2  4.0  140   9200     55    39  274      50            0  

[5 rows x 34 columns]

In [78]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
numeric=pd.DataFrame(scaler.fit_transform(numeric),columns=numeric.columns)
numeric.head()

Age    Weight    Length       BMI        DM       HTN  Current Smoker  \
0 -0.568446  1.351012  1.104333  0.522817 -0.650027  0.832308        1.951800   
1  0.780937 -0.320172 -0.828604  0.281122 -0.650027  0.832308       -0.512348   
2 -0.472061 -1.657120 -0.076906 -1.752404 -0.650027 -1.201478        1.951800   
3  0.684552 -0.570850 -0.721218 -0.100118 -0.650027  0.832308       -0.512348   
4 -0.857599  1.100334 -1.258145  2.423417 -0.650027  0.832308       -0.512348   

   EX-Smoker        FH        BP  ...       ESR        HB         K        Na  \
0  -0.184742 -0.433861 -1.034254  ... -0.783276  1.521673  1.025931  0.000868   
1  -0.184742 -0.433861  0.552475  ...  0.410930  0.464323  1.025931  3.946579   
2  -0.184742 -0.433861 -1.563163  ... -0.594717  0.215534  1.025931 -0.525227   
3  -0.184742 -0.433861 -1.563163  ...  3.553576 -0.655225  0.370114  0.263916   
4  -0.184742 -0.433861 -1.034254  ...  0.473783  0.028943 -0.504308 -0.262179   

        WBC     Lymph      Neut       PLT    EF-TTE  Region RWMA  
0 -0.772712  0.662975 -0.801571  0.650977  0.310686    -0.548761  
1  0.057248  0.562534 -0.506461 -0.930681 -0.811339     2.988994  
2 -0.067246  0.562534 -0.014609  0.140233 -0.811339     1.220117  
3  2.256644 -1.446280  1.165833  8.575744  0.871699    -0.548761  
4  0.679719  2.270027 -2.080384  0.865160  0.310686    -0.548761  

[5 rows x 34 columns]

In [79]:
X=pd.concat([objects,numeric],axis=1)
X.head()

Sex  Obesity  CRF  CVA  Airway disease  Thyroid Disease  CHF  DLP  \
0    1        1    0    0               0                0    0    1   
1    0        1    0    0               0                0    0    0   
2    1        0    0    0               0                0    0    0   
3    0        1    0    0               0                0    0    0   
4    0        1    0    0               0                0    0    0   

   Weak Peripheral Pulse  Lung rales  ...       ESR        HB         K  \
0                      0           0  ... -0.783276  1.521673  1.025931   
1                      0           0  ...  0.410930  0.464323  1.025931   
2                      0           0  ... -0.594717  0.215534  1.025931   
3                      0           0  ...  3.553576 -0.655225  0.370114   
4                      0           0  ...  0.473783  0.028943 -0.504308   

         Na       WBC     Lymph      Neut       PLT    EF-TTE  Region RWMA  
0  0.000868 -0.772712  0.662975 -0.801571  0.650977  0.310686    -0.548761  
1  3.946579  0.057248  0.562534 -0.506461 -0.930681 -0.811339     2.988994  
2 -0.525227 -0.067246  0.562534 -0.014609  0.140233 -0.811339     1.220117  
3  0.263916  2.256644 -1.446280  1.165833  8.575744  0.871699    -0.548761  
4 -0.262179  0.679719  2.270027 -2.080384  0.865160  0.310686    -0.548761  

[5 rows x 55 columns]

In [80]:
le2=LabelEncoder()
y=pd.DataFrame(le.fit_transform(y),columns=y.columns)
y.head()

/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/utils/validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


LAD
0    1
1    1
2    1
3    0
4    0

In [81]:
import numpy as np
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

test = SelectKBest(score_func=f_classif, k="all")
fit = test.fit(X, np.asarray(y).reshape(y.shape[0],))
features = fit.transform(X)
set_printoptions(precision=3)

scores=fit.scores_
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False)
scores

/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/feature_selection/_univariate_selection.py:115: UserWarning: Features [15] are constant.
  UserWarning)
/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/feature_selection/_univariate_selection.py:116: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


scores
33  80.590128
54  45.485196
13  42.897929
53  33.092610
21  31.677227
14  15.304161
38  14.080471
36  10.750653
50  10.716580
26  10.426517
31  10.054521
51   9.365250
45   9.201537
35   8.878939
37   6.982790
30   6.773142
39   6.133718
49   5.386946
11   5.052385
25   4.659874
47   4.150667
18   3.564077
24   3.315351
48   3.220880
32   3.206318
44   3.129196
34   3.119668
41   2.976623
0    2.891842
5    2.632829
9    2.578343
40   2.543350
27   2.230687
12   2.171572
28   1.983182
52   1.741406
2    1.562283
22   1.484608
16   1.430495
1    1.161460
46   0.949375
19   0.723033
6    0.711184
3    0.706746
7    0.681315
43   0.553019
42   0.482348
20   0.432956
23   0.264900
4    0.127271
10   0.127109
29   0.093403
17   0.021978
8    0.005218
15        NaN

In [82]:
scores=scores.fillna(0)

In [83]:
X_scores=scores[scores["scores"]>6]
len(X_scores)

17

In [84]:
Xp=X.iloc[:,X_scores.index]
Xp.head()

Typical Chest Pain  Region RWMA  Atypical    EF-TTE       Age  Nonanginal  \
0           -1.086212    -0.548761         0  0.310686 -0.568446           0   
1            0.920631     2.988994         0 -0.811339  0.780937           0   
2            0.920631     1.220117         0 -0.811339 -0.472061           0   
3           -1.086212    -0.548761         0  0.871699  0.684552           1   
4           -1.086212    -0.548761         0  0.310686 -0.857599           0   

   Tinversion  St Elevation     Lymph       HTN        PR      Neut       ESR  \
0    1.538397     -0.220097  0.662975  0.832308  0.546031 -0.801571 -0.783276   
1    1.538397     -0.220097  0.562534  0.832308  0.546031 -0.506461  0.410930   
2   -0.650027     -0.220097  0.562534 -1.201478  2.793959 -0.014609 -0.594717   
3   -0.650027     -0.220097 -1.446280  0.832308  0.546031  1.165833  3.553576   
4   -0.650027     -0.220097  2.270027  0.832308  0.546031 -2.080384  0.473783   

     Q Wave  St Depression        BP       FBS  
0 -0.236113       1.807652 -1.034254 -0.561315  
1 -0.236113       1.807652  0.552475 -0.753646  
2 -0.236113      -0.553204 -1.563163 -0.657481  
3 -0.236113       1.807652 -1.563163 -0.792113  
4 -0.236113      -0.553204 -1.034254 -0.292051

In [85]:
from sklearn import model_selection
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

In [86]:
from sklearn import svm
model1=svm.SVC(kernel="sigmoid",gamma="auto",random_state=0)
scores=cross_val_score(model1,Xp,np.asarray(y).reshape(y.shape[0],),cv=7)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.7888100271821202 +/- 0.04517436029259725


In [87]:
y_pred = cross_val_predict(model1, Xp,np.asarray(y).reshape(y.shape[0],) , cv=7)
AUC = roc_auc_score(y, y_pred)
print("AUC: ",AUC)

AUC:  0.776903416733925


In [88]:
print("Confusion matrix: \n",confusion_matrix(y,y_pred))

Confusion matrix: 
 [[ 89  37]
 [ 27 150]]


In [89]:
y.groupby("LAD")["LAD"].count()

LAD
0    126
1    177
Name: LAD, dtype: int64

In [90]:
print("Specificity: ",89/126)
print("Sensitivity: ",150/177)

Specificity:  0.7063492063492064
Sensitivity:  0.847457627118644


In [91]:
from sklearn.ensemble import RandomForestClassifier
model2=RandomForestClassifier(n_estimators=200,criterion="entropy",random_state=0)
scores=cross_val_score(model2,Xp,np.asarray(y).reshape(y.shape[0],),cv=7)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.7920567804288734 +/- 0.036980469152989456


In [92]:
y_pred = cross_val_predict(model2, Xp,np.asarray(y).reshape(y.shape[0],) , cv=7)
AUC = roc_auc_score(y, y_pred)
print("AUC: ",AUC)

AUC:  0.7751546946462201


In [27]:
print("Confusion matrix: \n",confusion_matrix(y,y_pred))

Confusion matrix: 
 [[ 85  41]
 [ 22 155]]


In [28]:
print("Specificity: ",85/126)
print("Sensitivity: ",155/177)

Specificity:  0.6746031746031746
Sensitivity:  0.8757062146892656


In [122]:
y=df[["LCX"]]
y.groupby("LCX")["LCX"].count()

LCX
Normal      184
Stenotic    119
Name: LCX, dtype: int64

In [123]:
y.head()

LCX
0    Normal
1  Stenotic
2    Normal
3    Normal
4    Normal

In [124]:
y=pd.DataFrame(le.fit_transform(y),columns=y.columns)
y.head()

/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/utils/validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


LCX
0    0
1    1
2    0
3    0
4    0

In [125]:
test = SelectKBest(score_func=f_classif, k="all")
fit = test.fit(X, np.asarray(y).reshape(y.shape[0],))
features = fit.transform(X)
set_printoptions(precision=3)

scores=fit.scores_
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False)
scores

/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/feature_selection/_univariate_selection.py:115: UserWarning: Features [15] are constant.
  UserWarning)
/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/feature_selection/_univariate_selection.py:116: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


scores
21  32.788936
33  28.229017
30  10.055553
40   9.652420
26   9.458768
13   8.835112
25   7.662383
44   6.850165
39   6.339915
52   5.107422
53   4.815632
54   4.437002
28   4.122313
41   3.984822
47   3.688393
0    3.545486
48   3.166124
14   2.993253
4    2.848583
34   2.581580
51   2.172157
2    1.925404
5    1.875258
23   1.716929
50   1.464735
38   1.432451
45   1.379504
24   1.308734
10   1.135029
11   1.127565
18   1.027680
27   0.888309
42   0.819138
35   0.811664
3    0.788711
49   0.761835
12   0.734962
6    0.645981
32   0.599959
1    0.535727
31   0.419460
37   0.343294
20   0.253419
43   0.227257
9    0.181728
46   0.109536
16   0.096510
36   0.078519
7    0.057461
22   0.055159
17   0.004702
29   0.002274
8    0.001116
19   0.000262
15        NaN

In [126]:
scores=scores.fillna(0)

In [127]:
X_scores=scores[scores["scores"]>1]
len(X_scores)

31

In [128]:
Xp=X.iloc[:,X_scores.index]
Xp.head()

Age  Typical Chest Pain        BP        CR       HTN  Atypical  \
0 -0.568446           -1.086212 -1.034254 -1.347725  0.832308         0   
1  0.780937            0.920631  0.552475 -0.210758  0.832308         0   
2 -0.472061            0.920631 -1.563163 -0.210758 -1.201478         0   
3  0.684552           -1.086212 -1.563163  0.547220  0.832308         0   
4 -0.857599           -1.086212 -1.034254 -0.210758  0.832308         0   

         DM       BUN       FBS       PLT  ...  CRF  Thyroid Disease  \
0 -0.650027 -1.368075 -0.561315  0.650977  ...    0                0   
1 -0.650027  1.799548 -0.753646 -0.930681  ...    0                0   
2 -0.650027 -0.072229 -0.657481  0.140233  ...    0                0   
3 -0.650027  1.799548 -0.792113  8.575744  ...    0                0   
4 -0.650027 -0.216212 -0.292051  0.865160  ...    0                0   

     Length     Lymph  Tinversion       ESR       BMI  Systolic Murmur  \
0  1.104333  0.662975    1.538397 -0.783276  0.522817                0   
1 -0.828604  0.562534    1.538397  0.410930  0.281122                0   
2 -0.076906  0.562534   -0.650027 -0.594717 -1.752404                0   
3 -0.721218 -1.446280   -0.650027  3.553576 -0.100118                0   
4 -1.258145  2.270027   -0.650027  0.473783  2.423417                1   

   Diastolic Murmur  Poor R Progression  
0                 0                   0  
1                 0                   0  
2                 0                   0  
3                 1                   0  
4                 0                   0  

[5 rows x 31 columns]

In [129]:
model3=svm.SVC(random_state=0)
scores=cross_val_score(model3,Xp,np.asarray(y).reshape(y.shape[0],),cv=5)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.6830054644808743 +/- 0.02616134129426984


In [130]:
y_pred = cross_val_predict(model3, Xp,np.asarray(y).reshape(y.shape[0],) , cv=5)
AUC = roc_auc_score(y, y_pred)
print("AUC: ",AUC)

AUC:  0.6441359152356595


In [131]:
print("Confusion matrix: \n",confusion_matrix(y,y_pred))

Confusion matrix: 
 [[152  32]
 [ 64  55]]


In [132]:
print("Specificity: ",152/(152+32))
print("Sensitivity: ",55/(55+64))

Specificity:  0.8260869565217391
Sensitivity:  0.46218487394957986


In [133]:
model4=RandomForestClassifier(n_estimators=90,criterion="entropy",random_state=0)
scores=cross_val_score(model4,Xp,np.asarray(y).reshape(y.shape[0],),cv=8)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.6931009957325747 +/- 0.028798534624506187


In [134]:
y_pred = cross_val_predict(model4, Xp,np.asarray(y).reshape(y.shape[0],) , cv=8)
AUC = roc_auc_score(y, y_pred)
print("AUC: ",AUC)

AUC:  0.6463509316770186


In [135]:
print("Confusion matrix: \n",confusion_matrix(y,y_pred))

Confusion matrix: 
 [[159  25]
 [ 68  51]]


In [154]:
print("Specificity: ",159/(159+25))
print("Sensitivity: ",51/(51+68))

Specificity:  0.8641304347826086
Sensitivity:  0.42857142857142855


In [156]:
y=df[["RCA"]]
y.groupby("RCA")["RCA"].count()

RCA
Normal      189
Stenotic    114
Name: RCA, dtype: int64

In [157]:
y.head()

RCA
0  Stenotic
1    Normal
2    Normal
3    Normal
4    Normal

In [158]:
y=pd.DataFrame(le2.fit_transform(y),columns=y.columns)
y.head()

/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/utils/validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


RCA
0    1
1    0
2    0
3    0
4    0

In [159]:
test = SelectKBest(score_func=f_classif, k="all")
fit = test.fit(X, np.asarray(y).reshape(y.shape[0],))
features = fit.transform(X)
set_printoptions(precision=3)

scores=fit.scores_
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False)
scores

/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/feature_selection/_univariate_selection.py:115: UserWarning: Features [15] are constant.
  UserWarning)
/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/feature_selection/_univariate_selection.py:116: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


scores
33  27.882997
25  23.749423
13  17.783381
51  16.592680
21  15.659448
50  14.619286
18  10.684980
45  10.579119
14   7.206624
26   6.704392
39   6.660558
41   5.836566
34   5.023049
49   4.913157
53   4.532393
0    4.491651
35   4.490805
47   3.601229
12   3.147255
40   2.864402
11   2.785396
23   2.553003
30   2.438596
48   2.399360
54   1.935916
6    1.661526
31   1.336244
16   1.211203
38   1.150348
3    1.081360
8    1.081360
52   1.043573
36   0.954702
28   0.671991
22   0.597491
7    0.491695
27   0.448242
37   0.407857
44   0.392974
4    0.296574
43   0.258480
24   0.251281
5    0.248755
10   0.242903
46   0.109718
29   0.092735
32   0.086474
42   0.082057
17   0.062412
2    0.047703
20   0.013761
1    0.009852
9    0.007672
19   0.002857
15        NaN

In [160]:
scores=scores.fillna(0)

In [161]:
X_scores=scores[scores["scores"]>2]
len(X_scores)

24

In [162]:
Xp=X.iloc[:,X_scores.index]
Xp.head()

Typical Chest Pain        DM  Atypical      Neut       Age     Lymph  \
0           -1.086212 -0.650027         0 -0.801571 -0.568446  0.662975   
1            0.920631 -0.650027         0 -0.506461  0.780937  0.562534   
2            0.920631 -0.650027         0 -0.014609 -0.472061  0.562534   
3           -1.086212 -0.650027         0  1.165833  0.684552 -1.446280   
4           -1.086212 -0.650027         0 -2.080384 -0.857599  2.270027   

   Poor R Progression       ESR  Nonanginal       HTN  ...    EF-TTE  Sex  \
0                   0 -0.783276           0  0.832308  ...  0.310686    1   
1                   0  0.410930           0  0.832308  ... -0.811339    0   
2                   0 -0.594717           0 -1.201478  ... -0.811339    1   
3                   0  3.553576           1  0.832308  ...  0.871699    0   
4                   0  0.473783           0  0.832308  ...  0.310686    0   

     Q Wave         K  Dyspnea        CR  Diastolic Murmur    Length  \
0 -0.236113  1.025931        0 -1.347725                 0  1.104333   
1 -0.236113  1.025931        0 -0.210758                 0 -0.828604   
2 -0.236113  1.025931        0 -0.210758                 0 -0.076906   
3 -0.236113  0.370114        1  0.547220                 1 -0.721218   
4 -0.236113 -0.504308        1 -0.210758                 0 -1.258145   

         BP        Na  
0 -1.034254  0.000868  
1  0.552475  3.946579  
2 -1.563163 -0.525227  
3 -1.563163  0.263916  
4 -1.034254 -0.262179  

[5 rows x 24 columns]

In [167]:
model5=svm.SVC(kernel="linear",probability=True,random_state=0)
scores=cross_val_score(model5,Xp,np.asarray(y).reshape(y.shape[0],),cv=5)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.7127868852459016 +/- 0.031186534885992517


In [164]:
y_pred = cross_val_predict(model5, Xp,np.asarray(y).reshape(y.shape[0],) , cv=5)
AUC = roc_auc_score(y, y_pred)
print("AUC: ",AUC)

AUC:  0.6862990810359232


In [165]:
print("Confusion matrix: \n",confusion_matrix(y,y_pred))

Confusion matrix: 
 [[150  39]
 [ 48  66]]


In [53]:
model6=RandomForestClassifier(n_estimators=90,random_state=0)
scores=cross_val_score(model6,Xp,np.asarray(y).reshape(y.shape[0],),cv=5)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.6800546448087431 +/- 0.05038772894785792


In [54]:
from sklearn.naive_bayes import BernoulliNB
model7=BernoulliNB()
scores=cross_val_score(model7,Xp,np.asarray(y).reshape(y.shape[0],),cv=5)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.7097267759562842 +/- 0.029185443975481798


In [55]:
from sklearn.ensemble import GradientBoostingClassifier
model8=GradientBoostingClassifier(learning_rate=0.015,max_features="sqrt",max_depth=4,
                                  random_state=0)
scores=cross_val_score(model8,Xp,np.asarray(y).reshape(y.shape[0],),cv=5)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.6997267759562842 +/- 0.021040735074453767


#### What we have so far:

- For "LAD" (Left Anterior Descending): 79.2% +/- 3.7%
- For "LCX" (Left Circumflex): 69.3% +/- 2.6%
- For "RCA" (Right Coronary Artery): 71.3% +/- 3.1%

Let's optimize the AUC or accuracy of the more promising models.

### LAD

In [93]:
y_prob =pd.DataFrame((cross_val_predict(model2, X, np.asarray(y).reshape(y.shape[0],), cv=7,method="predict_proba"))).rename(columns={0:"pred_0",1:"pred_1"})
y_prob.head()

pred_0  pred_1
0   0.580   0.420
1   0.115   0.885
2   0.090   0.910
3   0.445   0.555
4   0.670   0.330

In [94]:
y_probs1=y_prob[["pred_1"]]

In [95]:
import numpy as np
thresholds=np.arange(0,1,0.001)

In [96]:
def to_labels(pos_probs,threshold):
    return (pos_probs >= threshold).astype('int')

In [120]:
scores=[accuracy_score(y, to_labels(y_probs1,t)) for t in thresholds]

In [121]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.47000000000000003, 0.7920792079207921)


Accuracy is already at 79.2%.

In [103]:
scores=[roc_auc_score(y, to_labels(y_probs1,t)) for t in thresholds]

In [104]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.5650000000000001, 0.7820823244552058)


The original AUC score was 0.775.

In [105]:
len(y_probs1[y_probs1["pred_1"]>=0.565])

177

In [106]:
y_prob["pred_0"]=np.where((y_prob.pred_0 >=0.565),1,y_prob.pred_0)
y_prob["pred_0"]=np.where((y_prob.pred_0 < 0.565),0,y_prob.pred_0)
y_prob["pred_1"]=np.where((y_prob.pred_1 >=0.565),1,y_prob.pred_1)
y_prob["pred_1"]=np.where((y_prob.pred_1 <0.565),0,y_prob.pred_1)

In [107]:
len11=len(y_prob[y_prob["pred_1"]==1])
print("The number of predicted positive cases: ",len11)

The number of predicted positive cases:  177


In [110]:
y_prob["pred_0"]=np.where((y_prob.pred_0 == y_prob.pred_1),1,y_prob.pred_0)

In [111]:
len00=len(y_prob[y_prob["pred_0"]==1])
print("The number of predicted negative cases: ",len00)

The number of predicted negative cases:  126


In [114]:
from sklearn.preprocessing import OneHotEncoder
enc=OneHotEncoder(handle_unknown="ignore")
y_true=pd.DataFrame(enc.fit_transform(y).toarray())

In [116]:
test_pred=enc.inverse_transform(y_prob)

In [117]:
print("Accuracy after adjusting the threshold: ",accuracy_score(y,test_pred))

Accuracy after adjusting the threshold:  0.7755775577557755


In [118]:
print("Confusion matrix for test data after adjusting the threshold: \n",confusion_matrix(y,test_pred))

Confusion matrix for test data after adjusting the threshold: 
 [[ 92  34]
 [ 34 143]]


Adjusting the threshold resulted in more correct classifications of normal patients, but fewer correct classifications for patients with LAD risk. Since the side effects of at-risk patients as normal is greater, the increase in AUC score gained through adjusting the threshold is not worthwhile.

### LCX

In [136]:
y_prob =pd.DataFrame((cross_val_predict(model4, X, np.asarray(y).reshape(y.shape[0],), cv=8,method="predict_proba"))).rename(columns={0:"pred_0",1:"pred_1"})
y_prob.head()

pred_0    pred_1
0  0.700000  0.300000
1  0.344444  0.655556
2  0.744444  0.255556
3  0.655556  0.344444
4  0.811111  0.188889

In [137]:
y_probs1=y_prob[["pred_1"]]

In [138]:
import numpy as np
thresholds=np.arange(0,1,0.001)

In [139]:
def to_labels(pos_probs,threshold):
    return (pos_probs >= threshold).astype('int')

In [140]:
scores=[accuracy_score(y, to_labels(y_probs1,t)) for t in thresholds]

In [141]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.434, 0.6897689768976898)


No more room to improve accuracy through adjusting the threshold.

In [142]:
scores=[roc_auc_score(y, to_labels(y_probs1,t)) for t in thresholds]

In [143]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.367, 0.6755800146145415)


Current AUC for "LCX" is 0.646.

In [144]:
len(y_probs1[y_probs1["pred_1"]>=0.367])

162

In [145]:
y_prob["pred_0"]=np.where((y_prob.pred_0 >=0.367),1,y_prob.pred_0)
y_prob["pred_0"]=np.where((y_prob.pred_0 < 0.367),0,y_prob.pred_0)
y_prob["pred_1"]=np.where((y_prob.pred_1 >=0.367),1,y_prob.pred_1)
y_prob["pred_1"]=np.where((y_prob.pred_1 <0.367),0,y_prob.pred_1)

In [146]:
len11=len(y_prob[y_prob["pred_1"]==1])
print("The number of predicted positive cases: ",len11)

The number of predicted positive cases:  162


In [149]:
y_prob["pred_0"]=np.where((y_prob.pred_0 == y_prob.pred_1),0,y_prob.pred_0)

In [150]:
len00=len(y_prob[y_prob["pred_0"]==1])
print("The number of predicted negative cases: ",len00)

The number of predicted negative cases:  141


In [151]:
test_pred=enc.inverse_transform(y_prob)

In [152]:
print("Accuracy after adjusting the threshold: ",accuracy_score(y,test_pred))

Accuracy after adjusting the threshold:  0.6600660066006601


In [153]:
print("Confusion matrix for test data after adjusting the threshold: \n",confusion_matrix(y,test_pred))

Confusion matrix for test data after adjusting the threshold: 
 [[111  73]
 [ 30  89]]


In [155]:
print("Specificity: ", 111/(111+73))
print("Sensitivity: ",89/(89+30))

Specificity:  0.6032608695652174
Sensitivity:  0.7478991596638656


For "LCX", adjusting the threshold resulted in accuracy that is 3.3% lower. However, sensitivity to at-risk patients was increased by more than 30%. Since the model is now able to detect more at-risk patients, the 3.3% decrease in overall accuracy is worthwile.

#### RCA

In [168]:
y_prob =pd.DataFrame((cross_val_predict(model5, X, np.asarray(y).reshape(y.shape[0],), cv=5,method="predict_proba"))).rename(columns={0:"pred_0",1:"pred_1"})
y_prob.head()

pred_0    pred_1
0  0.742183  0.257817
1  0.676421  0.323579
2  0.664489  0.335511
3  0.772229  0.227771
4  0.685081  0.314919

In [169]:
y_probs1=y_prob[["pred_1"]]

In [170]:
import numpy as np
thresholds=np.arange(0,1,0.001)

In [171]:
def to_labels(pos_probs,threshold):
    return (pos_probs >= threshold).astype('int')

In [173]:
scores=[accuracy_score(y, to_labels(y_probs1,t)) for t in thresholds]

In [174]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.429, 0.66996699669967)


In [175]:
scores=[roc_auc_score(y, to_labels(y_probs1,t)) for t in thresholds]

In [176]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.33, 0.676065162907268)


For "RCA", no more room for improvement through adjusting the threshold.

#### Final summary:

|CAD (Coronary Artery Disease) cause|Model|Cross validation folds|Accuracy|Standard deviation|AUC|Specificity|Sensitivity|
|-----------------------------------|-----|----------------------|--------|------------------|---|-----------|-----------|
|LAD (Left Anterior Descending)|RandomForestClassifier| 7-fold cross validation|79.2%|3.7%|0.775|67.5%|87.6%|
|LCX (Left Circumflex), before adjusting the threshold|RandomForestClassifier|8-fold cross validation|69.3%|2.9%|0.64|86%|43%|
|LCX (Left Circumflex), after adjusting the threshold|RandomForestClassifier|8-fold cross validation|66%| | 0.678|60.3%|74.8%|
|RCA (Right Coronary Artery)|Support Vector Machine, "linear" kernel|5-fold cross validation|71.3%|3.1%|0.686|79.4%|57.9%|